tests/Client.hs

```diff
@@ -0,0 +1,232 @@
+              M.pkcResponse =
+                M.AuthenticatorAttestationResponse
+                  { M.arcClientData =
+                      M.CollectedClientData
+                        { M.ccdChallenge = pkcocChallenge,
+                          M.ccdOrigin = M.Origin "https://localhost:8080",
+                          M.ccdCrossOrigin = Just True,
+                          M.ccdHash = M.ClientDataHash $ hashlazy clientDataBS
+                        },
+                    M.arcAttestationObject = attestationObject,
+                    M.arcTransports = Set.fromList [M.AuthenticatorTransportUSB, M.AuthenticatorTransportBLE, M.AuthenticatorTransportNFC, M.AuthenticatorTransportInternal]
+                  },
+              M.pkcClientExtensionResults = M.AuthenticationExtensionsClientOutputs {}
+            }
+  pure (response, authenticator')
+
+-- | Performs assertion as per the client specification provided an
+-- authenticator. MonadRandom is required for signing using Ed25519 which
+-- requires a random number to be generated during signing. There exists
+-- methods to not rely on a random number, but these have not been implemented
+-- in the cryptonite library we rely on.
+clientAssertion :: (MonadFail m, MonadRandom m) => JS.PublicKeyCredentialRequestOptions -> Authenticator -> m (JS.RequestedPublicKeyCredential, Authenticator)
+clientAssertion options authenticator = do
+  let Right M.PublicKeyCredentialRequestOptions {..} = decodePublicKeyCredentialRequestOptions options
+      clientDataBS =
+        encode
+          JS.ClientDataJSON
+            { JS.typ = "webauthn.get",
+              JS.challenge = decodeUtf8 . Base64.encode $ M.unChallenge pkcogChallenge,
+              JS.origin = "https://localhost:8080/",
+              JS.crossOrigin = Nothing
+            }
+      clientDataHash = M.ClientDataHash $ hashlazy clientDataBS
+      -- See https://www.w3.org/TR/webauthn-2/#sctn-discover-from-external-source step 17.3.isEmpty
+      allowCredentialDescriptorList = case pkcogAllowCredentials of
+        [] -> Nothing
+        xs -> Just xs
+  ((credentialId, authenticatorData, signature, userHandle), authenticator') <-
+    authenticatorGetAssertion
+      authenticator
+      (fromMaybe (M.RpId "localhost") pkcogRpId)
+      clientDataHash
+      allowCredentialDescriptorList
+      True
+      True
+      pkcogExtensions
+  let response =
+        encodeRequestedPublicKeyCredential
+          M.PublicKeyCredential
+            { M.pkcIdentifier = credentialId,
+              M.pkcResponse =
+                M.AuthenticatorAssertionResponse
+                  { M.argClientData =
+                      M.CollectedClientData
+                        { M.ccdChallenge = pkcogChallenge,
+                          M.ccdOrigin = M.Origin "https://localhost:8080",
+                          M.ccdCrossOrigin = Just True,
```

Comment:
Same here. Also shouldn't this directly be linked to the `JS.CollectedClientData`?

Reply:
I'm unsure what you mean with "directly linked".

Reply:
In both `clientAttestation` and `clientAssertion`, both `M.CollectedClientData` and `JS.CollectedClientData` are declared. The latter should arguably be created by encoding the former.

Reply:
I was confused as we don't have a type called `JS.CollectedClientData` but I see what you mean now, I'll implement it!